### PR TITLE
Support grouping by computed keys

### LINKED ISSUE #535: Let groupby use arbitrary expressions
The `by ...` clause in a groupby proc is currently a list of field names.  We should generalize this to allow an arbitrary expression.  This would enable lots of things that can't be easily expressed right now, such as
```
count() by len(someset)
count() by String.tolower(something)
count() by 1000*Math.round(bytes/1000)
```

### PR DESCRIPTION
**Pull request by henridf**
closes #535

### CODE REVIEW

File: proc/groupby.go [hunk 1]
```diff
@@ -15,12 +15,13 @@ import (
 	"github.com/brimsec/zq/zcode"
 	"github.com/brimsec/zq/zng"
 	"github.com/brimsec/zq/zng/resolver"
+	"github.com/pkg/errors"
```
Comment (alfred-landrum) [inline proc/groupby.go:18]: Was this intentional, or did you mean just `"errors"`?

Comment (henridf) [reply proc/groupby.go:18]: oops, it was certainly not! i should pay closer attention to what goimports is doing for me :)

Comment (alfred-landrum) [reply proc/groupby.go]: I figured, the same has happened to me :)

File: ast/ast.go [hunk 2]
```diff
@@ -283,11 +284,11 @@ type (
 
 	PutProc struct {
 		Node
-		Clauses []PutClause `json:"clauses"`
+		Clauses []Assignment `json:"clauses"`
```
Comment (nwt) [inline ast/ast.go:287]: If you're going to rename this type, please finish the job by renaming the fields and variables of the type, either in this PR or in another. Leaving the clause terminology in place is confusing.

Comment (henridf) [reply ast/ast.go:287]: Ok. How far would you like this to go: all the way including renaming things like  https://github.com/brimsec/zq/blob/master/proc/put.go#L39 ?

Comment (henridf) [reply ast/ast.go:287]: (I'll do this in a separate PR upon answer to the above question).

**Pull request merged**